Want better error message for attempted hygiene violation, binding in macro

### Code

```Rust
macro_rules! let_it { {} => { let it = (); } }
macro_rules! print_it { {} => { println!("{:?}", it); } }

fn main() {
    let_it!();
    let () = it;
    print_it!();
}
```

### Current output

```Shell
error[E0425]: cannot find value `it` in this scope
 --> src/main.rs:6:14
  |
6 |     let () = it;
  |              ^^ not found in this scope

error[E0425]: cannot find value `it` in this scope
 --> src/main.rs:2:50
  |
2 | macro_rules! print_it { {} => { println!("{:?}", it); } }
  |                                                  ^^ not found in this scope
...
7 |     print_it!();
  |     ----------- in this macro invocation
  |
  = note: this error originates in the macro `print_it` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Desired output

```Shell
error[E0425]: cannot find value `it` in this scope
 --> src/main.rs:6:14
  |
6 |     let () = it;
  |              ^^ not found in this scope

help: An identifier with the same name is defined here, but is not accessible due to macro hygiene
1 | macro_rules! let_it { {} => { let it = (); } }
  |                                   ^^ inaccessible definition 

error[E0425]: cannot find value `it` in this scope
 --> src/main.rs:2:50
  |
2 | macro_rules! print_it { {} => { println!("{:?}", it); } }
  |                                                  ^^ not found in this scope
...
7 |     print_it!();
  |     ----------- in this macro invocation
  |

help: An identifier with the same name is defined here, but is not accessible due to macro hygiene
1 | macro_rules! let_it { {} => { let it = (); } }
  |                                   ^^ inaccessible definition 

  = note: this error originates in the macro `print_it` (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Rationale and extra context

See #148580

### Other cases

See #148580 in which I first suggested such a hint.  The example I gave there  https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=1ee109560f0d9e2f216a02e346679f1c  binds outside the macro and tries to use from inside.  This was implemented in #148612 (thanks @chenyukang !)

However, here we have the *binding* in the macro, which is also forbidden by hygiene.  Apparently the implemnetation in #148612 doesn't cover this case.

### Rust Version

Today's nightly on the Rust playground.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=9fb1d5c7091fc5b8572aa30e9ae2f039

Also locally:

```Shell
rustc 1.93.0-nightly (646a3f8c1 2025-12-02)
binary: rustc
commit-hash: 646a3f8c15baefb98dc6e0c1c1ba3356db702d2a
commit-date: 2025-12-02
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.5
```

### Anything else?

CC @JonathanBrouwer 

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_resolve/src/late/diagnostics.rs
+++ b/compiler/rustc_resolve/src/late/diagnostics.rs
@@ -1155,6 +1155,7 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
         let callsite_span = span.source_callsite();
         for rib in self.ribs[ValueNS].iter().rev() {
             for (binding_ident, _) in &rib.bindings {
+                // Case 1: the identifier is defined in the same scope as the macro is called
                 if binding_ident.name == ident.name
                     && !binding_ident.span.eq_ctxt(span)
                     && !binding_ident.span.from_expansion()
@@ -1163,6 +1164,19 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                     err.span_help(
                         binding_ident.span,
                         "an identifier with the same name exists, but is not accessible due to macro hygiene",
+                    );
+                    return;
+                }
+
+                // Case 2: the identifier is defined in a macro call in the same scope
+                if binding_ident.name == ident.name
+                    && binding_ident.span.from_expansion()
+                    && binding_ident.span.source_callsite().eq_ctxt(callsite_span)
+                    && binding_ident.span.source_callsite().lo() < callsite_span.lo()
+                {
+                    err.span_help(
+                        binding_ident.span,
+                        "an identifier with the same name is defined here, but is not accessible due to macro hygiene",
                     );
                     return;
                 }

--- a/tests/ui/hygiene/pattern-macro.stderr
+++ b/tests/ui/hygiene/pattern-macro.stderr
@@ -3,6 +3,16 @@ error[E0425]: cannot find value `x` in this scope
    |
 LL |     x + 1;
    |     ^ not found in this scope
+   |
+help: an identifier with the same name is defined here, but is not accessible due to macro hygiene
+  --> $DIR/pattern-macro.rs:1:28
+   |
+LL | macro_rules! foo { () => ( x ) }
+   |                            ^
+...
+LL |     let foo!() = 2;
+   |         ------ in this macro invocation
+   = note: this error originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)
 
 error: aborting due to 1 previous error
 

--- a/tests/ui/macros/macro-hygiene-help-issue-149604.rs
+++ b/tests/ui/macros/macro-hygiene-help-issue-149604.rs
@@ -1,0 +1,9 @@
+macro_rules! let_it { {} => { let it = (); } }
+macro_rules! print_it { {} => { println!("{:?}", it); } }
+//~^ ERROR cannot find value `it` in this scope
+
+fn main() {
+    let_it!();
+    let () = it; //~ ERROR cannot find value `it` in this scope
+    print_it!();
+}

--- a/tests/ui/macros/macro-hygiene-help-issue-149604.stderr
+++ b/tests/ui/macros/macro-hygiene-help-issue-149604.stderr
@@ -1,0 +1,38 @@
+error[E0425]: cannot find value `it` in this scope
+  --> $DIR/macro-hygiene-help-issue-149604.rs:7:14
+   |
+LL |     let () = it;
+   |              ^^ not found in this scope
+   |
+help: an identifier with the same name is defined here, but is not accessible due to macro hygiene
+  --> $DIR/macro-hygiene-help-issue-149604.rs:1:35
+   |
+LL | macro_rules! let_it { {} => { let it = (); } }
+   |                                   ^^
+...
+LL |     let_it!();
+   |     --------- in this macro invocation
+   = note: this error originates in the macro `let_it` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0425]: cannot find value `it` in this scope
+  --> $DIR/macro-hygiene-help-issue-149604.rs:2:50
+   |
+LL | macro_rules! print_it { {} => { println!("{:?}", it); } }
+   |                                                  ^^ not found in this scope
+...
+LL |     print_it!();
+   |     ----------- in this macro invocation
+   |
+help: an identifier with the same name is defined here, but is not accessible due to macro hygiene
+  --> $DIR/macro-hygiene-help-issue-149604.rs:1:35
+   |
+LL | macro_rules! let_it { {} => { let it = (); } }
+   |                                   ^^
+...
+LL |     let_it!();
+   |     --------- in this macro invocation
+   = note: this error originates in the macro `print_it` which comes from the expansion of the macro `let_it` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0425`.

--- a/tests/ui/proc-macro/gen-macro-rules-hygiene.stderr
+++ b/tests/ui/proc-macro/gen-macro-rules-hygiene.stderr
@@ -30,6 +30,16 @@ error[E0425]: cannot find value `local_def` in this scope
    |
 LL |         local_def;
    |         ^^^^^^^^^ help: a local variable with a similar name exists: `local_use`
+   |
+help: an identifier with the same name is defined here, but is not accessible due to macro hygiene
+  --> $DIR/gen-macro-rules-hygiene.rs:13:1
+   |
+LL | gen_macro_rules!();
+   | ^^^^^^^^^^^^^^^^^^
+...
+LL |         generated!();
+   |         ------------ in this macro invocation
+   = note: this error originates in the macro `generated` (in Nightly builds, run with -Z macro-backtrace for more info)
 
 error: aborting due to 3 previous errors
 

--- a/tests/ui/proc-macro/mixed-site-span.stderr
+++ b/tests/ui/proc-macro/mixed-site-span.stderr
@@ -606,6 +606,13 @@ error[E0425]: cannot find value `local_def` in this scope
    |
 LL |         local_def;
    |         ^^^^^^^^^ help: a local variable with a similar name exists: `local_use`
+   |
+help: an identifier with the same name is defined here, but is not accessible due to macro hygiene
+  --> $DIR/mixed-site-span.rs:23:9
+   |
+LL |         proc_macro_rules!();
+   |         ^^^^^^^^^^^^^^^^^^^
+   = note: this error originates in the macro `proc_macro_rules` (in Nightly builds, run with -Z macro-backtrace for more info)
 
 error: aborting due to 52 previous errors
 

